### PR TITLE
osd: a faster version of SnapMapper::get_prefix()

### PR DESCRIPTION
**Pull request by ronen-fr**
Using fmtlib, and treating the size of uint64_t as a known
quantity.
Per google-bench - the new version is 10% to 20% faster
(tested on my laptop).

Signed-off-by: Ronen Friedman <rfriedma@redhat.com>

### CODE REVIEW

File: src/osd/SnapMapper.cc [hunk 1]
```diff
@@ -99,13 +100,11 @@ int OSDriver::get_next(
...
 string SnapMapper::get_prefix(int64_t pool, snapid_t snap)
 {
-  char buf[100];
-  int len = snprintf(
-    buf, sizeof(buf),
-    "%lld_%.*X_",
-    (long long)pool,
-    (int)(sizeof(snap)*2), static_cast<unsigned>(snap));
-  return MAPPING_PREFIX + string(buf, len);
+  assert(sizeof(snap) == 8); // 2*8 is the '16' in the '%.16X' format below
```
Comment (rzarzynski) [inline src/osd/SnapMapper.cc:103]: Hmm, the sake of consistency suggests `ceph_assert()`. I guess this shouldn't inflate the check's penalty as (IIRC; please check me) we already define `assert` to be an alias to `ceph_assert`.

Comment (rzarzynski) [reply src/osd/SnapMapper.cc:103]: ^H^H^H

How about `static_assert` instead?

Comment (ronen-fr) [reply src/osd/SnapMapper.cc:103]: Yes. It should have been static_assert. Will fix.

**Comment by ronen-fr** [comment]
[rfriedma-2022-08-02_16:18:28-rados:thrash-wip-rf-snprefix-distro-default-smithi](http://pulpito.front.sepia.ceph.com/rfriedma-2022-08-02_16:18:28-rados:thrash-wip-rf-snprefix-distro-default-smithi/#collapseOne) - 45 rados:thrash tests OK

**Review by tchaikov — APPROVED** [review]

**Review by rzarzynski — APPROVED** [review]

**Comment by athanatos** [comment]
I wrote some unit tests for the existing key types to ensure we don't change them by accident: https://github.com/ceph/ceph/commit/dd2808c62e7b9f4ece45a835a183e0893ead8d60 -- probably worth lumping it in with the PR.

LGTM otherwise -- definitely cleaner.

**Comment by ronen-fr** [comment]
> Might be worth to also update SnapMapper's `make_purged_snap_key` and `get_prefix`?

Yes - but originally I wanted to get this merged first. Now - as will be adding the tests - will commit those too.

File: src/test/test_snap_mapper.cc [hunk 2]
```diff
@@ -668,6 +678,82 @@ TEST_F(SnapMapperTest, MultiPG) {
...
+	    std::numeric_limits<int64_t>::max() - 20, "namespace"},
+	  "OBJ_.1_7FFFFFFFFFFFFFEB.76543210.ffffffffffffffec.test%uobject..namespace"}
+    });
+
+  for (auto &[object, test_object_key]: object_to_object_key) {
+    auto object_key = get_tester().to_object_key(object);
+    std::cout << object << " should be "
+	      << test_object_key << " is "
+	      << get_tester().to_object_key(object)
+	      << std::endl;
```
Comment (tchaikov) [inline src/test/test_snap_mapper.cc:703]: IMHO, a successful run should be quiet.

Comment (ronen-fr) [reply src/test/test_snap_mapper.cc:703]: Agree. I will remove the log lines

Comment (athanatos) [reply src/test/test_snap_mapper.cc:706]: Well, it should probably print in the case of a failure.

**Comment by ronen-fr** [comment]
(missed the opportunity to name this PR "a faster shade of snap")

**Comment by ronen-fr** [comment]
jenkins retest this please

**Comment by ronen-fr** [comment]
jenkins retest api

**Comment by ronen-fr** [comment]
jenkins test api

**Review by Matan-B — APPROVED** [review]

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2022-08-07_05:39:26-rados-wip-kefu-testing-2022-08-07-1123-distro-default-smithi/
- https://pulpito.ceph.com/kchai-2022-08-07_07:44:14-rados-wip-kefu-testing-2022-08-07-1123-distro-default-smithi/

the `sudo yum -y install ceph-radosgw` failures are tracked by https://tracker.ceph.com/issues/57051

**Pull request merged**